Gracefully handle large messages.

mreid-moz opened the pull request:
Now that max_message_size is a configuration parameter, there is a
risk that the sandbox decoder can attempt to output messages that
are too large, resulting in an error like:

FATAL: process_message() heka/sandbox/decoders/http_edge_decoder.lua:98: inject_message() output_limit exceeded

Which takes down Heka. We now wrap `inject_message` with `pcall`
to ensure we trap errors.

mreid-moz committed:
Gracefully handle large messages.

Now that max_message_size is a configuration parameter, there is a
risk that the sandbox decoder can attempt to output messages that
are too large, resulting in an error like:

FATAL: process_message() heka/sandbox/decoders/http_edge_decoder.lua:98: inject_message() output_limit exceeded

Which takes down Heka. We now wrap `inject_message` with `pcall`
to ensure we trap errors.

trink commented:
The patch file has not been removed and the build script has not been updated

trink commented on heka/sandbox/decoders/http_edge_decoder.lua:
```diff
@@ -173,6 +174,19 @@ function process_message()
+    local main_status, main_err = pcall(inject_message, main_msg)
+    local error_message = ""
+    local return_code = 0
+    if not landfill_status then
```
Why is this not tested right after the injection?

mreid-moz replied:
The idea was to try both "inject" operations to minimize any data loss, so the error reporting is delayed until the end of processing.

trink replied:
Yeah, that is not the way it works. The failure will discard all messages so even in main was successful it will never see the light of day.

mreid-moz replied:
D'oh. I'll change the code to signal failure right away then.

trink commented on heka/sandbox/decoders/http_edge_decoder.lua:
```diff
@@ -173,6 +174,19 @@ function process_message()
+        return_code = -1
+    end
+
+    if return_code == 0 then return return_code end
```
Just initialize error_message to nil and you can remove this line

mreid-moz commented:
The patch/build change is in PR #18

mreid-moz committed:
Don't try to inject main_msg if landfill_msg fails

Per @trink, the failure will discard all messages so even if one
was successful it will never see the light of day.

mreid-moz committed:
Merge branch 'master' into large_message_resiliency

Conflicts:
	examples/basic_local_pipeline.toml
